packages/core/core/src/requests/AssetGraphRequest.js

```diff
@@ -0,0 +1,251 @@
+import AssetGraph from '../AssetGraph';
+import {PARCEL_VERSION} from '../constants';
+import createEntryRequest from './EntryRequest';
+import createTargetRequest from './TargetRequest';
+import createAssetRequest from './AssetRequest';
+import createPathRequest from './PathRequest';
+
+import dumpToGraphViz from '../dumpGraphToGraphViz';
+
+type AssetGraphRequestInput = {|
+  entries?: Array<string>,
+  assetGroups?: Array<AssetGroup>,
+  optionsRef: SharedReference,
+  name: string,
+|};
+
+type RunInput = {|
+  input: AssetGraphRequestInput,
+  ...StaticRunOpts<{|
+    assetGraph: AssetGraph,
+    changedAssets: Map<string, Asset>,
+    assetRequests: Array<AssetGroup>,
+  |}>,
+|};
+
+type AssetGraphRequest = {|
+  id: string,
+  +type: 'asset_graph_request',
+  run: RunInput => Async<{|
+    assetGraph: AssetGraph,
+    changedAssets: Map<string, Asset>,
+    assetRequests: Array<AssetGroup>,
+  |}>,
+  input: AssetGraphRequestInput,
+|};
+
+export default function createAssetGraphRequest(
+  input: AssetGraphRequestInput,
+): AssetGraphRequest {
+  return {
+    type: 'asset_graph_request',
+    id: input.name,
+    run: input => {
+      let builder = new AssetGraphBuilder(input);
+      return builder.build();
+    },
+    input,
+  };
+}
+
+const typesWithRequests = new Set([
+  'entry_specifier',
+  'entry_file',
+  'dependency',
+  'asset_group',
+]);
+
+export class AssetGraphBuilder {
+  assetGraph: AssetGraph;
+  assetRequests: Array<AssetGroup>;
+  queue: PromiseQueue<mixed>;
+  changedAssets: Map<string, Asset> = new Map();
+  optionsRef: SharedReference;
+  api: RunAPI;
+  name: string;
+  assetRequests: Array<AssetGroup> = [];
+  cacheKey: string;
+
+  constructor({input, prevResult, api, options}: RunInput) {
+    let {entries, assetGroups, optionsRef, name} = input;
+    let assetGraph = prevResult?.assetGraph ?? new AssetGraph();
+    assetGraph.setRootConnections({
+      entries,
+      assetGroups,
+    });
+    this.assetGraph = assetGraph;
+    this.optionsRef = optionsRef;
+    this.api = api;
+    this.name = name;
+
+    let {hot, publicUrl, distDir, minify, scopeHoist} = options;
+    this.cacheKey = md5FromObject({
+      parcelVersion: PARCEL_VERSION,
+      name,
+      options: {hot, publicUrl, distDir, minify, scopeHoist},
+      entries,
+    });
+
+    this.queue = new PromiseQueue();
+  }
+
+  async build(): Promise<{|
+    assetGraph: AssetGraph,
+    changedAssets: Map<string, Asset>,
+    assetRequests: Array<AssetGroup>,
+  |}> {
+    let errors = [];
+
+    let root = this.assetGraph.getRootNode();
+    if (!root) {
+      throw new Error('A root node is required to traverse');
+    }
+
+    let visited = new Set([root.id]);
+
+    const visit = node => {
+      if (errors.length > 0) {
+        return;
+      }
+
+      if (this.shouldSkipRequest(node)) {
+        visitChildren(node);
+      } else {
+        // ? do we need to visit children inside of the promise that is queued?
```

Comment:
As opposed to in parallel?

Reply:
I'm pretty sure my concern when I wrote this comment is that we only `await this.queue.run()` on line 152, so it might not be guaranteed that we visit the children before moving on to store and return the result?